packages/nodejs/src/storage/StorageBackendDiskAsync.ts

```diff
@@ -134,14 +135,40 @@ export class StorageBackendDiskAsync extends MaybeAsyncStorage {
             return this.#writeFile(fileName, value);
         }
 
-        const promise = writeFile(this.filePath(fileName), value, "utf8").finally(() => {
+        // TODO write temp, fsync, move
+        const promise = this.#writeAndMoveFile(this.filePath(fileName), value).finally(() => {
             this.#writeFileBlocker.delete(fileName);
         });
         this.#writeFileBlocker.set(fileName, promise);
 
         return promise;
     }
 
+    async #writeAndMoveFile(filepath: string, value: string): Promise<void> {
+        const tmpName = `${filepath}.tmp`;
+        await writeFile(tmpName, value, { encoding: "utf8", flush: true }); // flush does fsync after write
+        await rename(tmpName, filepath);
+    }
+
+    /**
+     * fsync on a directory ensures there are no rename operations etc. which haven't been persisted to disk.
+     * We do this as best effort to ensure that all writes are persisted to disk.
+     */
+    async #fsyncStorageDir() {
+        const fd = await open(this.#path, "r");
+        try {
+            await fd.sync();
+        } catch (error) {
+            // Windows will cause `EPERM: operation not permitted, fsync`
+            // for directories, so lets catch this generically
```

Comment:
Did you try on windows with "w" flag for opening?

Assuming that doesn't work, I think you could branch the code... For windows instead of opening directory, after rename open the new file in write mode, then sync that fd.

Reply:
Adjusted